src/Security/TwoFactor/Provider.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Security\TwoFactor;
+
+use Contao\User;
+use Scheb\TwoFactorBundle\Security\TwoFactor\AuthenticationContextInterface;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Provider\TwoFactorFormRendererInterface;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Provider\TwoFactorProviderInterface;
+
+class Provider implements TwoFactorProviderInterface
+{
+    /**
+     * @var Authenticator
+     */
+    private $authenticator;
+
+    /**
+     * @var TwoFactorFormRendererInterface
+     */
+    private $formRenderer;
+
+    /**
+     * @var bool
+     */
+    private $enforceTwoFactor;
+
+    /**
+     * @param Authenticator                  $authenticator
+     * @param TwoFactorFormRendererInterface $formRenderer
+     * @param bool                           $enforceTwoFactor
+     */
+    public function __construct(Authenticator $authenticator, TwoFactorFormRendererInterface $formRenderer, bool $enforceTwoFactor)
+    {
+        $this->authenticator = $authenticator;
+        $this->formRenderer = $formRenderer;
+        $this->enforceTwoFactor = $enforceTwoFactor;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function beginAuthentication(AuthenticationContextInterface $context): bool
+    {
+        $user = $context->getUser();
+
+        if (!$user instanceof User) {
+            return false;
+        }
+
+        if (!$user->secret) {
+            return false;
+        }
+
+        if (!$this->enforceTwoFactor && !$user->useTwoFactor) {
```

Comment:
So if the user does not have a secret (which is the default, if he never edited his data), then we don't enforce 2FA?

Reply:
This seems correct to me. Otherwise the user would be asked to enter their verification code although they never had a chance to set up 2FA.

Reply:
In fact that's wrong. If there is no secret and 2FA is enforced, a secret should be generated, so the user gets the QR code to verify on the login screen.

Reply:
Fixed in eac180e.